### PR TITLE
parser: add RestoreCtx and change the Node interface

### PR DESCRIPTION
**Pull request by leoppro**
add RestoreCtx to output in many formats

See: pingcap/tidb#8679

support flags for restore function, we can control the format of output sql text by flags

Update proposal pr: pingcap/tidb#8694

### About `go.mod1`:
we have to update code(`parser` and `tidb`) at the same time when we change the `Node` interface  because parser and tidb are cyclic dependency.
in order to compile successfully, i have to replace `pingcap/tidb` with `leoppro/tidb` temporarily (and also repale `pingcap/parser` with `leoppro/parser` in tidb).
i will follow this to handle the dependence problem:
1. merge this pr
2. change go.mod of pingcap/tidb#8679, update dependency `pingcap/parser` to master
3. merge pingcap/tidb#8679
4. commit a pr to pingcap/parser,change dependency tidb to master

### CODE REVIEW

**Comment by leoppro** [comment]
@tiancaiamao @XuHuaiyu  PTAL

File: ast/util.go [hunk 1]
```diff
@@ -62,14 +66,97 @@ func (checker *readOnlyChecker) Leave(in Node) (out Node, ok bool) {
...
+
+// WriteKeyWord write the keyword into writer
+func (ctx *RestoreCtx) WriteKeyWord(keyWord string) {
+	switch {
+	case ctx.Flags.Has(RestoreKeyWordUppercase):
+		keyWord = strings.ToUpper(keyWord)
+	case ctx.Flags.Has(RestoreKeyWordLowercase):
+		keyWord = strings.ToLower(keyWord)
+	}
+	_, _ = fmt.Fprint(ctx.In, keyWord)
```
Comment (GregoryIan) [inline ast/util.go]: remove `_, _ =`

File: ast/util.go [hunk 2]
```diff
@@ -62,14 +66,97 @@ func (checker *readOnlyChecker) Leave(in Node) (out Node, ok bool) {
...
+	RestoreKeyWordUppercase
+	RestoreKeyWordLowercase
+
+	RestoreNameUppercase
+	RestoreNameLowercase
+	RestoreNameOriginal
+	RestoreNameSingleQuotes
+	RestoreNameDoubleQuotes
+	RestoreNameBackQuote
+	RestoreNameEscapeBackQuote
```
[the remark this answers is not in the document]
Comment (leoppro) [reply ast/util.go]: got it

File: ast/util.go [hunk 3]
```diff
@@ -62,14 +66,97 @@ func (checker *readOnlyChecker) Leave(in Node) (out Node, ok bool) {
...
+// str maybe wrapped in quotes and escape according to RestoreFlags
+func (ctx *RestoreCtx) WriteString(str string) {
+	if ctx.Flags.Has(RestoreStringEscapeBackslash) {
+		str = strings.Replace(str, `\`, `\\`, -1)
+	}
+	quotes := ""
+	switch {
+	case ctx.Flags.Has(RestoreStringSingleQuotes):
+		str = strings.Replace(str, `'`, `''`, -1)
+		quotes = `''`
```
Comment (kennytm) [inline ast/util.go]:
```suggestion
		quotes = `'`
```

File: ast/util.go [hunk 4]
```diff
@@ -62,14 +66,97 @@ func (checker *readOnlyChecker) Leave(in Node) (out Node, ok bool) {
...
+		name = strings.Replace(name, "`", "``", -1)
+		quotes = "`"
+	}
+	fmt.Fprint(ctx.In, quotes, name, quotes)
 }
 
-// EscapeName escape the `name`
-func EscapeName(name string) string {
-	return strings.Replace(name, "`", "``", -1)
+// WriteName write the plain text into writer without any handling
```
Comment (kennytm) [inline ast/util.go]:
```suggestion
// WritePlain write the plain text into writer without any handling
```

**Comment by leoppro** [comment]
Thank you for your careful review. @kennytm

**Comment by XuHuaiyu** [comment]
Please add an explanation in the description for why we make this change.

Or

Update the proposal for this change.

**Comment by leoppro** [comment]
i will update the proposal @XuHuaiyu

File: go.mod1 [hunk 5]
```diff
@@ -15,3 +15,5 @@ require (
 	golang.org/x/net v0.0.0-20181029044818-c44066c5c816
 	golang.org/x/text v0.3.0
 )
+
+replace github.com/pingcap/tidb => github.com/leoppro/tidb v0.0.0-00000000000000-4efb42782bbc7b9589ebf4afb2f82416c7a76fa6
```
Comment (tiancaiamao) [inline go.mod1]: Why this change ?

Comment (leoppro) [reply go.mod1]: we can't compile successfully if we didn't change the go.mod because we changed the Node interface.

Comment (leoppro) [reply go.mod1]: after the pr (pingcap/tidb#8679) merged, we can change it back.

**Comment by tiancaiamao** [comment]
LGTM

File: ast/util.go [hunk 6]
```diff
@@ -62,14 +66,102 @@ func (checker *readOnlyChecker) Leave(in Node) (out Node, ok bool) {
...
+	RestoreNameBackQuote
+)
+
+const (
+	DefaultRestoreFlags = RestoreStringSingleQuotes | RestoreKeyWordUppercase | RestoreNameOriginal |
+		RestoreNameBackQuote
+)
+
+// Has return weather `rf` has this flag
+func (rf RestoreFlags) Has(flag RestoreFlags) bool {
```
Comment (XuHuaiyu) [inline ast/util.go]: I think we'd better split this function into small ones.
e.g. parser/mysql/types.go HasXXXFlag() bool

**Comment by leoppro** [comment]
@XuHuaiyu PTAL again

File: ast/util.go [hunk 7]
```diff
@@ -79,89 +85,132 @@ const (
...
+func (rf RestoreFlags) HasStringSingleQuotesFlag() bool {
+	return rf.Has(RestoreStringSingleQuotes)
+}
+
+// HasStringDoubleQuotesFlag returns weather `rf` has `RestoreStringDoubleQuotes` flag.
+func (rf RestoreFlags) HasStringDoubleQuotesFlag() bool {
+	return rf.Has(RestoreStringDoubleQuotes)
+}
+
+// HasStringEscapeBackslashFlag returns weather `rf` has `RestoreStringEscapeBackslash` flag.
```
Comment (XuHuaiyu) [inline ast/util.go]: s/ weather/ whether

File: opcode/opcode.go [hunk 8]
```diff
@@ -136,3 +138,12 @@ var opsLiteral = map[Op]string{
...
 	fmt.Fprintf(w, "%s", opsLiteral[o])
 }
+
+// Restore the Op into a Writer
+func (o Op) Restore(w io.Writer) error {
+	if v, ok := opsLiteral[o]; ok {
+		fmt.Fprint(w, v)
+		return nil
+	}
+	return errors.Errorf("Invalid opcode type: %d", o)
```
Comment (XuHuaiyu) [inline opcode/opcode.go]: refine this error msg.
